src/System.Memory/tests/ReadOnlyBuffer/ReadableBufferFacts.cs

```diff
@@ -0,0 +1,220 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Linq;
+using Xunit;
+
+namespace System.Memory.Tests
+{
+    public abstract class ReadOnlyBufferFacts
+    {
+        public class Array: ReadOnlyBufferFacts
+        {
+            public Array() : base(ReadOnlyBufferFactory.Array) { }
+        }
+
+        public class OwnedMemory: ReadOnlyBufferFacts
+        {
+            public OwnedMemory() : base(ReadOnlyBufferFactory.OwnedMemory) { }
+        }
+
+        public class SingleSegment: ReadOnlyBufferFacts
+        {
+            public SingleSegment() : base(ReadOnlyBufferFactory.SingleSegment) { }
+        }
+
+        public class SegmentPerByte: ReadOnlyBufferFacts
+        {
+            public SegmentPerByte() : base(ReadOnlyBufferFactory.SegmentPerByte) { }
+        }
+
+        internal ReadOnlyBufferFactory Factory { get; }
+
+        internal ReadOnlyBufferFacts(ReadOnlyBufferFactory factory)
+        {
+            Factory = factory;
+        }
+
+        [Fact]
+        public void EmptyIsCorrect()
+        {
+            var buffer = Factory.CreateOfSize(0);
+            Assert.Equal(0, buffer.Length);
+            Assert.True(buffer.IsEmpty);
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(8)]
+        public void LengthIsCorrect(int length)
+        {
+            var buffer = Factory.CreateOfSize(length);
+            Assert.Equal(length, buffer.Length);
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(8)]
+        public void ToArrayIsCorrect(int length)
```

Comment:
Do we wount to add outerloop tests for very long lengths? i.e. longer than int.

Reply:
What might it show? It's all linear, so it will just make things linearly slower.

Reply:
I was thinking it would be good to have tests for buffers longer than int.maxvalue to ensure that we don't have bugs related to int length overflows.

Reply:
https://github.com/dotnet/corefx/issues/26768